src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -271,35 +267,40 @@ public void AddToPending(ChainedHeaderBlock chainedHeaderBlock)
         {
             this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock.ChainedHeader);
 
-            this.blocksQueue.Enqueue(chainedHeaderBlock);
+            lock (this.getBlockLock)
+            {
+                this.batch.Add(chainedHeaderBlock);
+                this.currentBatchSizeBytes += chainedHeaderBlock.Block.BlockSize.Value;
+            }
 
             this.logger.LogTrace("(-)");
         }
 
         /// <summary>
-        /// Dequeues the blocks continuously and saves them to the database when max batch size is reached or timer ran out.
+        /// Periodically checks whether the batch of blocks is full enough to be saved to the database,
+        /// or if sufficient time elapses save the batch anyway.
         /// </summary>
         /// <remarks>Batch is always saved on shutdown.</remarks>
-        private async Task DequeueBlocksContinuouslyAsync()
+        private async Task CheckBatchContinuouslyAsync()
         {
             this.logger.LogTrace("()");
 
-            Task<ChainedHeaderBlock> dequeueTask = null;
-            Task timerTask = null;
+            int cumulativeTicks = 0;
+            bool saveBatch = false;
 
             while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
             {
-                // Start new dequeue task if not started already.
-                dequeueTask = dequeueTask ?? this.blocksQueue.DequeueAsync();
-
-                // Wait for one of the tasks: dequeue or timer (if available) to finish.
-                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
-
-                bool saveBatch = false;
-
                 try
                 {
-                    await task.ConfigureAwait(false);
+                    await Task.Delay(1000);
+
+                    cumulativeTicks++;
+
+                    if (cumulativeTicks >= BatchMaxSaveIntervalSeconds)
+                    {
+                        saveBatch = true;
+                        cumulativeTicks = 0;
```

Comment:
reset this when the batch was saved. If you do it here cumulativeTicks won't be reset in case we saved the batch because it was too big and not because of the timer. That will lead to additional save when we don't need it.